src/_posts/2019-10-30-crafting-websites-using-finite-state-machines-part-I.md

```diff
@@ -0,0 +1,403 @@
+
+{:context {:status "list"
+          :filter true}
+:todos [{:text "todo 1" :done? true}
+        {:text "todo 2" :done? false}]}
+
+
+;What about edit?
+;How do we know which do we edit?
+;How do we order the todos?
+
+{:context {:status "edit"
+          :filter true
+          :selected-id 1},
+:todos {
+        :1 {:id 1, :text "todo 1", :done? true, :time 1}
+        :2 {:id 2, :text "todo 2", :done? false, :time 2}
+        }}
+
+;Let's carve it in stone:
+
+(require '[clojure.spec.alpha :as s])
+
+;SPEC
+(s/def ::status #{"list" "edit" "add"})
+(s/def ::filter #(or (false? %) (true? %)))
+(s/def ::filter-with-nil #(or (nil? %) (false? %) (true? %)))
+(s/def ::selected-id nat-int?)
+(s/def ::context (s/keys :req-un [::status] :opt-un [::filter-id ::selected-id]))
+
+
+(s/def ::id nat-int?)
+(s/def ::text string?)
+(s/def ::done? boolean?)
+(s/def ::time number?)
+(s/def ::todo (s/keys :req-un [::id ::text ::done? ::time]))
+(s/def ::todos (s/map-of keyword? ::todo))
+
+(defn edit-mode-has-a-valid-selected-id? [state]
+  (if (= "edit" (get-in state [:context :status]))
+    (some (into #{} (map :id (vals (:todos state)))) [(get-in state [:context :selected-id])])
+    true
+    ))
+
+(s/def ::model (s/and
+                 (s/keys :req-un [::context] :opt-un [::todos])
+                 edit-mode-has-a-valid-selected-id?
+                 ))
+
+(comment
+(s/valid? ::model {:context {:status "edit"
+          :filter true
+          :selected-id 1}
+:todos {
+        :1 {:id 1, :text "todo 1", :done? true, :time 1}
+        :2 {:id 2, :text "todo 2", :done? false, :time 2}
+        }}))
+```
+
+### Step 3: Solve the data problem
+
+
+Now we know how we could represent our states as data in the model, let's test drive the different transitions which will al be functions in the controller. Normally I start with a plan, where I will know what I want to test.
+
+The plan of the tests:
+
+ * [ ] controller-should
+ * [ ] initialize-in-list-mode
+ * [ ] check&amp;uncheck
+ * [ ] delete
+ * [ ] toggle-filters
+ * [ ] set a filter
+ * [ ] remove all filters
+ * [ ] set-add-mode
+ * [ ] save-a-new-todo
+ * [ ] set-list-mode
+ * [ ] set-one-for-edit
+ * [ ] save-a-changed-todo
+
+Then we'll write the first one:
+
+```clojure
+(deftest initialize-in-list-mode
+         (is (= {:context {:status "list"} :todos   {}}
+                (init!))))
+```
+
+which will obviously
+
+Test fails! Good... now we write the code to make it pass:
+
+```clojure
+;THE MODEL
+(def model (atom {:context {:status "list"}}))
+
+
+;THE CONTROLLER
+(defn commit! [value atom-to-change]
+  (reset! atom-to-change value))
+
+(defn init! []
+  (-> @model
+    (assoc :todos {})
+    (assoc-in [:context :status] "list")
+    (commit! model)))
+
+
+
+;THE TEST
+(deftest initialize-in-list-mode
+         (is (= {:context {:status "list"} :todos {}}
+                (init!))))
+```
+
+[![TDD](https://img.youtube.com/vi/CvEoXuoAhCg/0.jpg)](https://www.youtube.com/watch?v=CvEoXuoAhCg)
```

Comment:
I think something needs to be improved here, as it is not clear it is a video. I say just embed the video as normal, rather than this way.

Reply:
Removed the video, doesn't add value.